jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -40,23 +50,24 @@
  */
 public class StepsRunner {
 
-  /** Holds the individual steps. */
-  private static class Steps {
+  /** Holds the individual step results. */
+  private static class StepResults {
 
-    @Nullable private RetrieveRegistryCredentialsStep retrieveTargetRegistryCredentialsStep;
-    @Nullable private AuthenticatePushStep authenticatePushStep;
-    @Nullable private PullBaseImageStep pullBaseImageStep;
-    @Nullable private PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
-
-    @Nullable
-    private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
-
-    @Nullable private PushLayersStep pushBaseImageLayersStep;
-    @Nullable private PushLayersStep pushApplicationLayersStep;
-    @Nullable private BuildImageStep buildImageStep;
-    @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
+    private static <E> Future<E> failedFuture() {
+      return Futures.immediateFailedFuture(
+          new IllegalStateException("invalid usage; required step not configured"));
+    }
 
-    @Nullable private AsyncStep<BuildResult> finalStep;
+    private Future<ImageAndAuthorization> baseImageAndAuth = failedFuture();
+    private Future<List<Future<CachedLayerAndName>>> baseImageLayers = failedFuture();
+    private List<Future<CachedLayerAndName>> applicationLayers = new ArrayList<>();
```

Comment:
why is this not a `failedFuture()` ?

Reply:
About why it is not a `Future` at the top level: in both of the original and new designs, you already know what (and how many) application layers you should build at the beginning, so it is different from `baseImageLayers` where you don't know what and how many layers to download.

I think I'll just make it `@Nullable` and do `checkNotNull` to guard against a wrong schedule order as in the original design. (To be frank, I instantiated only to avoid having to use `checkNotNull`, but on second thought, it's better to explicitly guard wrong usage.)